File: catalog/backends/python-django/patterns/rag_chat_mongo_route.py
import os
import json
from openai import OpenAI
from pymongo import MongoClient
from django.http import JsonResponse
from ddtrace.llmobs.decorators import workflow, embedding, retrieval, llm

# Pattern: LLM Obs — RAG chat using MongoDB vector search
# Adapt: replace system prompt and retrieval logic with domain context
MONGO_URL = os.environ.get("MONGODB_URL", "mongodb://demo:demo@mongodb:27017/demo")
_oai = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))
_db = MongoClient(MONGO_URL).get_default_database()


@embedding(model_name="text-embedding-3-small", model_provider="openai")
def _embed_query(text):
    resp = _oai.embeddings.create(model="text-embedding-3-small", input=[text])
    return resp.data[0].embedding


@retrieval
def _search_docs(query_vec):
    results = _db.documents.aggregate([
        {"$vectorSearch": {"index": "vector_index", "path": "embedding", "queryVector": query_vec, "numCandidates": 50, "limit": 3}},
        {"$project": {"title": 1, "content": 1, "score": {"$meta": "vectorSearchScore"}}},
    ])
    return [{"title": r["title"], "content": r["content"], "score": r["score"]} for r in results]


@llm(model_name="gpt-4o-mini", model_provider="openai")
def _generate(messages):
    resp = _oai.chat.completions.create(model="gpt-4o-mini", messages=messages)
    return resp.choices[0].message.content


@workflow
def chat(request):
    body = json.loads(request.body)
    user_msg = body.get("message", "")
    history = body.get("history", [])
    query_vec = _embed_query(user_msg)
    sources = _search_docs(query_vec)
    context = "\n\n".join(f"[{s['title']}]: {s['content']}" for s in sources)
    messages = [
        {"role": "system", "content": f"Answer using this context:\n{context}"},
        *history,
        {"role": "user", "content": user_msg},
    ]
    answer = _generate(messages)
    return JsonResponse({"response": answer, "sources": [{"title": s["title"], "score": s["score"]} for s in sources]})

File: catalog/backends/node-express/patterns/rag_chat_mongo_route.js
const { MongoClient } = require('mongodb');
const { OpenAI } = require('openai');
const tracer = require('dd-trace');

// Pattern: LLM Obs — RAG chat using MongoDB vector search
// Adapt: replace system prompt and retrieval logic with domain context
const client = new MongoClient(process.env.MONGODB_URL || 'mongodb://demo:demo@mongodb:27017/demo');
const db = client.db();
const oai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
const llmobs = tracer.llmobs;

const _embedQuery = llmobs.wrap({ kind: 'embedding', modelName: 'text-embedding-3-small', modelProvider: 'openai' },
  async function embedQuery(text) {
    const resp = await oai.embeddings.create({ model: 'text-embedding-3-small', input: [text] });
    return resp.data[0].embedding;
  }
);

const _searchDocs = llmobs.wrap({ kind: 'retrieval' },
  async function searchDocs(queryVec) {
    return db.collection('documents').aggregate([
      { $vectorSearch: { index: 'vector_index', path: 'embedding', queryVector: queryVec, numCandidates: 50, limit: 3 } },
      { $project: { title: 1, content: 1, score: { $meta: 'vectorSearchScore' } } },
    ]).toArray();
  }
);

const _generate = llmobs.wrap({ kind: 'llm', modelName: 'gpt-4o-mini', modelProvider: 'openai' },
  async function generate(messages) {
    const resp = await oai.chat.completions.create({ model: 'gpt-4o-mini', messages });
    return resp.choices[0].message.content;
  }
);

const chat = llmobs.wrap({ kind: 'workflow', name: 'rag-chat-mongo' },
  async function ragChat(req, res) {
    const { message, history = [] } = req.body;
    const queryVec = await _embedQuery(message);
    const sources = await _searchDocs(queryVec);
    const context = sources.map(s => `[${s.title}]: ${s.content}`).join('\n\n');
    const messages = [
      { role: 'system', content: `Answer using this context:\n${context}` },
      ...history,
      { role: 'user', content: message },
    ];
    const answer = await _generate(messages);
    res.json({ response: answer, sources: sources.map(s => ({ title: s.title, score: s.score })) });
  }
);

module.exports = { chat };
